### PR TITLE
implemented Bi-LSTM Model 

### PR DESCRIPTION
**Pull request by floschne**
with accuracy, precision, recall and f1 evaluation metrics

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -1,51 +1,89 @@
...
-        net = tf.layers.dense(net, units=units, activation=tf.nn.relu)
-    logits = tf.layers.dense(net, units=params['n_classes'], activation=None)
-    predicted_classes = tf.argmax(logits, axis=1)
     if mode == tf.estimator.ModeKeys.PREDICT:
+        # Predictions
         predictions = {
-            'class_ids' : predicted_classes[:, tf.newaxis],
-            'probabilities' : tf.nn.softmax(logits),
-            'logits' : logits}
+            'preds': treshed_preds,
```
Comment (julianbetz) [inline src/model.py]: Probabilities for class 1 vs. class 0 can still be provided.
```suggestion
            'class_ids': tf.cast(treshed_preds, tf.int64),
            'probabilities': preds
```

File: src/model.py [hunk 2]
```diff
@@ -1,51 +1,89 @@
...
+        }
         return tf.estimator.EstimatorSpec(mode, predictions=predictions)
-    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
-    accuracy = tf.metrics.accuracy(labels=labels, predictions=predicted_classes, name='acc_op')
-    metrics = {'accuracy' : accuracy}
-    tf.summary.scalar('accuracy', accuracy[1])
+
+    # Loss function is binary (sigmoid) cross entropy
+    loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=labels, logits=logits[:, :, 0]))
```
Comment (julianbetz) [inline src/model.py]: We need to ensure loss = 0 for all entries that correspond to padding. Did you check that? Maybe we should apply the seq_len_mask here as well before aggregating.

reduce_mean then operates on differently sized sequences, i.e. they should be properly weighted.

File: src/model.py [hunk 3]
```diff
@@ -1,51 +1,89 @@
...
+        tf.summary.scalar(metric_name, op[1])
+
     if mode == tf.estimator.ModeKeys.EVAL:
         return tf.estimator.EstimatorSpec(mode, loss=loss, eval_metric_ops=metrics)
-    optimizer = tf.train.AdagradOptimizer(learning_rate=params['learning_rate'])
-    train_op = optimizer.minimize(loss, global_step=tf.train.get_global_step())
-    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)
+
+    elif mode == tf.estimator.ModeKeys.TRAIN:
+        train_op = tf.train.AdamOptimizer().minimize(loss, global_step=tf.train.get_or_create_global_step())
```
Comment (julianbetz) [inline src/model.py:94]: If it is no longer used, we should probably discard the learning rate from the hyperparameter dictionary and use the Adam default.

File: src/main.py [hunk 4]
```diff
@@ -2,59 +2,89 @@
...
-N_FEATURES = [loader.load(loader.ids[0])[0].shape[1] for loader in [DataLoader(os.path.abspath(os.path.dirname(os.path.abspath(__file__)) + '/../dat/fast_load'), tst_size=0, seed=SEED)]][0]
+N_FEATURES = [loader.load(loader.ids[0])[0].shape[1] for loader in [
+    DataLoader(os.path.abspath(os.path.dirname(os.path.abspath(__file__)) + '/../dat/fast_load'), tst_size=0,
+               seed=SEED)]][0]
 FEATURE_COLS = [tf.feature_column.numeric_column(key='features', shape=N_FEATURES)]
-HYPERPARAMS = dict(learning_rate=hp.loguniform('learning_rate', -7, -3)) # TODO Add hyperparameters
+HYPERPARAMS = dict(learning_rate=hp.loguniform('learning_rate', -7, -3))  # TODO Add hyperparameters
+
+# Setup logging
+Path('results').mkdir(exist_ok=True)
```
Comment (julianbetz) [inline src/main.py]: We should probably make this path relative to the source file, not the working directory, in order to be able to collect results in the same directory.

File: src/main.py [hunk 5]
```diff
@@ -150,10 +190,11 @@ def convert(alignments, spectrograms):
             with open(utterances_output_dir + '/%s.json' % (key,), 'w') as file_handle:
                 # file_dict = {'id' : key, 'alignments' : value}
                 json.dump(value, file_handle)
-        progress.print_bar(i + 1, n_ids, 20, 'Storing alignment data... ┃', '┃ DONE %.4fs' % (time() - start_time))
-        print('Storing sequence IDs...', end='', flush=True)
-        with open(output_dir + '/utterances.json', 'w') as file_handle:
-            json.dump(sorted(tag_dict), file_handle, indent=4) # Asserts the same IDs in both alignments and spectrograms
+            progress.print_bar(i + 1, n_ids, 20, 'Storing alignment data... ┃', '┃ DONE %.4fs' % (time() - start_time))
```
Comment (julianbetz) [inline src/main.py]: Wrong indentation starting at this line

File: src/data_handler.py [hunk 6]
```diff
@@ -233,8 +241,8 @@ def load():
             elif mode == PRED_MODE:
                 yield {'features': feat_seq, 'length': length}
 
-    dtypes = ({'features' : tf.float32, 'length' : tf.int64}, tf.float32)
-    shapes = ({'features' : tf.TensorShape([None, N_FEATURES]), 'length' : tf.TensorShape([])}, tf.TensorShape([None]))
+    dtypes = ({'features': tf.float32, 'length': tf.int32}, tf.float32)
```
Comment (julianbetz) [inline src/data_handler.py:244]: Are you sure that this type change is in line with all other parts of the project code?

File: src/main.py [hunk 7]
```diff
@@ -2,59 +2,89 @@
...
+              help='The proportion of training data per fold in cross-validation. The remaining data of this fold is provided as evaluation data.',
+              show_default=True)
+@click.option('--batch_size', '-b', default=None, type=int,
+              help='The batch size for training. If not specified, the full training set is used.')
+@click.option('--n_epochs', '-e', default=1, help='The number of passes over the whole training set.',
+              show_default=True)
+@click.option('--max_hyperparam_sets', default=100,
+              help='The maximum number of hyperparameter sets to try during hyperparameter optimization.',
+              show_default=True)
+@click.option('--lstm_size', default=100, help='The size of the Bi-LSTM Cells')
```
Comment (julianbetz) [inline src/main.py]: I suppose that should be a hyperparameter.

File: src/main.py [hunk 8]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
         for i, (feat, id) in enumerate(zip(feats, ids)):
             progress.print_bar(i, n_ids, 20, 'Storing spectrogram data... ┃', '┃')
             np.save(utterances_output_dir + '/%s.npy' % (id,), feat)
-        progress.print_bar(i + 1, n_ids, 20, 'Storing spectrogram data... ┃', '┃ DONE %.4fs' % (time() - start_time))
```
Comment (julianbetz) [inline src/main.py]: The initial indentation was correct.

File: src/main.py [hunk 9]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
...
+            progress.print_bar(i + 1, n_ids, 20, 'Storing spectrogram data... ┃',
+                               '┃ DONE %.4fs' % (time() - start_time))
+
+
+def build_estimator(model_dir, fold_id, learning_rate, lstm_size, batch_size):
+    # Params # TODO click option for dropout or is this done by hyperopt?
+    params = {
+        'dropout': 0.5,
+        'epochs': 25,
```
Comment (julianbetz) [inline src/main.py]: Epochs have to be controlled outside the estimator for mid-training evaluation to work with our implementation.

File: src/main.py [hunk 10]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
         for i, (feat, id) in enumerate(zip(feats, ids)):
             progress.print_bar(i, n_ids, 20, 'Storing spectrogram data... ┃', '┃')
             np.save(utterances_output_dir + '/%s.npy' % (id,), feat)
-        progress.print_bar(i + 1, n_ids, 20, 'Storing spectrogram data... ┃', '┃ DONE %.4fs' % (time() - start_time))
+            progress.print_bar(i + 1, n_ids, 20, 'Storing spectrogram data... ┃',
+                               '┃ DONE %.4fs' % (time() - start_time))
+
+
+def build_estimator(model_dir, fold_id, learning_rate, lstm_size, batch_size):
+    # Params # TODO click option for dropout or is this done by hyperopt?
```
Comment (julianbetz) [inline src/main.py:214]: It usually should be done via hyperopt.

File: src/main.py [hunk 11]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
...
+    params = {
+        'dropout': 0.5,
+        'epochs': 25,
+        'buffer': 15000,
```
Comment (julianbetz) [inline src/main.py]: Is this actually used?

File: src/main.py [hunk 12]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
...
+        'lstm_size': lstm_size,
+        'learning_rate': learning_rate,
+        'batch_size': batch_size
```
Comment (julianbetz) [inline src/main.py]: What does the estimator itself need the batch size for? It should be controlled outside the estimator.

File: src/main.py [hunk 13]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
...
+        estimator = build_estimator(model_dir, i, learning_rate, lstm_size, batch_size)
+
         # TODO Before initial evaluation, make sure that a zero-global-step checkpoint exists
-        eval_result = estimator.evaluate(input_fn=lambda:input_fn(loader, evl_ids, mode=EVAL_MODE)) # Evaluation
+        eval_result = estimator.evaluate(input_fn=lambda: input_fn(loader, evl_ids, mode=EVAL_MODE))  # Evaluation
         for epoch in range(n_epochs):
-            estimator.train(input_fn=lambda:input_fn(loader, trn_ids, batch_size, random_state, mode=TRAIN_MODE), steps=None) # Training
-            eval_result = estimator.evaluate(input_fn=lambda:input_fn(loader, evl_ids, mode=EVAL_MODE)) # Evaluation
-        eval_result = estimator.evaluate(input_fn=lambda:input_fn(loader, val_ids, mode=EVAL_MODE)) # Validation TODO Is logged as evalutation
+            batch_size = len(trn_ids) if maximize_batch_size else batch_size
```
Comment (julianbetz) [inline src/main.py]: Why compute it over in every epoch?

File: src/main.py [hunk 14]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
...
+            estimator.train(input_fn=lambda: input_fn(loader, trn_ids, batch_size, random_state, mode=TRAIN_MODE),
+                            steps=None)  # Training
+            eval_result = estimator.evaluate(input_fn=lambda: input_fn(loader, evl_ids, mode=EVAL_MODE))  # Evaluation
+
+        eval_result = estimator.evaluate(
+            input_fn=lambda: input_fn(loader, val_ids, mode=EVAL_MODE))  # Validation TODO Is logged as evalutation
         loss += eval_result['loss']
-    loss /= n_splits
-    progress.print_bar(i + 1, n_splits, 20, 'Cross-validation: ┃', '┃ Loss: %f' % (loss,))
+        loss /= n_splits
```
Comment (julianbetz) [inline src/main.py]: Wrong indentation: Division here makes the loss computation wrong.

File: src/main.py [hunk 15]
```diff
@@ -165,50 +206,76 @@ def convert(alignments, spectrograms):
...
+        progress.print_bar(i + 1, n_splits, 20, 'Cross-validation: ┃', '┃ Loss: %f' % (loss,))
```
Comment (julianbetz) [inline src/main.py]: Wrong indentation: The last call to the progress bars always has to be done after termination of the loop.

File: src/model.py (hunk 1 above)
Comment (julianbetz) [reply src/model.py]: Or:
```suggestion
            'class_ids': tf.cast(treshed_preds, tf.bool),
            'probabilities': preds
```

**Review by julianbetz — CHANGES_REQUESTED** [review]
Let's talk about this, espc. hyperopt, before merging.
